From d8f8c027239fb2c080aa3eb69f32e7cb116f0ad9 Mon Sep 17 00:00:00 2001
From: Zihao Bo <zihao.bo@sophgo.com>
Date: Wed, 8 Apr 2026 21:18:30 +0800
Subject: [PATCH] feat(ddr): Add support for new DDR vendors and pinmux
 configurations

[description ]: Add new DDR vendors matching eFuse leakage definitions:
  - DDR_VENDOR_ESMT_N21_2G (0b01100) = ESMT 2Gb DDR3 21nm
  - DDR_VENDOR_ESMT_N19_4G (0b01101) = ESMT 4Gb DDR3 19nm (placeholder)
  - DDR_VENDOR_UNILC_N25_512M_DDR2 (0b01110) = UniIC 512Mb DDR2 25nm
  - DDR_VENDOR_UNILC_N25_1G (0b01111) = UniIC 1Gb DDR3 25nm
  - DDR_VENDOR_UNILC_N21_2G (0b10000) = UniIC 2Gb DDR3 21nm
  - DDR_VENDOR_NY_N20_1G (0b01011) = Nanya 1Gb DDR3 20nm
  Includes pinmux, pkg_type mapping, and vendor switch entries for
  both cv180x and cv181x platforms.
[root   cause]: NA
[JIRA      ID]: NA
[chip project]: cv181x
[side effects]: NA
[Affected *.ko/*.so/*.a] : NA

Change-Id: I9ece36e6a9ac942c07ce4c9d5b859f000d70bfd9
---
 drivers/ddr/cvx16_pinmux.c         | 77 +++++++++++++++++++++++++++++-
 drivers/ddr/ddr_pkg_info.c         |  2 +-
 include/drivers/ddr/ddr_pkg_info.h |  7 ++-
 3 files changed, 83 insertions(+), 3 deletions(-)

diff --git a/drivers/ddr/cvx16_pinmux.c b/drivers/ddr/cvx16_pinmux.c
index 76e27a7..be5aaf9 100644
--- a/drivers/ddr/cvx16_pinmux.c
+++ b/drivers/ddr/cvx16_pinmux.c
@@ -39,7 +39,7 @@ void cvx16_pinmux(void)
 		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x67513028);
 		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000004);
 		break;
-	case DDR_VENDOR_NY_1G:
+	case DDR_VENDOR_NY_N20_1G:
 		// DDR3_1G
 		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x08060F0B);
 		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x07040109);
@@ -129,6 +129,81 @@ void cvx16_pinmux(void)
 		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x76153280);
 		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000004);
 		break;
+	case DDR_VENDOR_UNILC_N25_1G:
+		// UnilC_DDR3_1G
+		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x08020B07);
+		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x00030905);
+		mmio_write_32(0x0008 + PHYD_BASE_ADDR, 0x04010006);
+		mmio_write_32(0x000C + PHYD_BASE_ADDR, 0x1300150C);
+		mmio_write_32(0x0010 + PHYD_BASE_ADDR, 0x16111214);
+		mmio_write_32(0x0014 + PHYD_BASE_ADDR, 0x00000A10);
+		mmio_write_32(0x0018 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x001C + PHYD_BASE_ADDR, 0x00000100);
+		mmio_write_32(0x0020 + PHYD_BASE_ADDR, 0x08137564);
+		mmio_write_32(0x0024 + PHYD_BASE_ADDR, 0x00000002);
+		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x76513820);
+		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000004);
+		break;
+	case DDR_VENDOR_UNILC_N21_2G:
+		// UnilC_DDR3_2G
+		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x08070D09);
+		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x0605020B);
+		mmio_write_32(0x0008 + PHYD_BASE_ADDR, 0x14040100);
+		mmio_write_32(0x000C + PHYD_BASE_ADDR, 0x1503000C);
+		mmio_write_32(0x0010 + PHYD_BASE_ADDR, 0x0A120013);
+		mmio_write_32(0x0014 + PHYD_BASE_ADDR, 0x00111016);
+		mmio_write_32(0x0018 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x001C + PHYD_BASE_ADDR, 0x00000100);
+		mmio_write_32(0x0020 + PHYD_BASE_ADDR, 0x82135764);
+		mmio_write_32(0x0024 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x76513208);
+		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000004);
+		break;
+	case DDR_VENDOR_ESMT_N21_2G:
+		// ESMT_N21_DDR3_2G
+		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x08070D09);
+		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x0605020B);
+		mmio_write_32(0x0008 + PHYD_BASE_ADDR, 0x14040100);
+		mmio_write_32(0x000C + PHYD_BASE_ADDR, 0x1503000C);
+		mmio_write_32(0x0010 + PHYD_BASE_ADDR, 0x16001213);
+		mmio_write_32(0x0014 + PHYD_BASE_ADDR, 0x0011100A);
+		mmio_write_32(0x0018 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x001C + PHYD_BASE_ADDR, 0x00000100);
+		mmio_write_32(0x0020 + PHYD_BASE_ADDR, 0x82315746);
+		mmio_write_32(0x0024 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x67513208);
+		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000004);
+		break;
+	case DDR_VENDOR_ESMT_N19_4G:
+		// ESMT_DDR3_4G_19nm
+		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x09050D07);
+		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x00080E02);
+		mmio_write_32(0x0008 + PHYD_BASE_ADDR, 0x0C060B01);
+		mmio_write_32(0x000C + PHYD_BASE_ADDR, 0x12031104);
+		mmio_write_32(0x0010 + PHYD_BASE_ADDR, 0x15001310);
+		mmio_write_32(0x0014 + PHYD_BASE_ADDR, 0x000A1614);
+		mmio_write_32(0x0018 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x001C + PHYD_BASE_ADDR, 0x00000100);
+		mmio_write_32(0x0020 + PHYD_BASE_ADDR, 0x31756024);
+		mmio_write_32(0x0024 + PHYD_BASE_ADDR, 0x00000008);
+		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x45172380);
+		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000006);
+		break;
+	case DDR_VENDOR_UNILC_N25_512M_DDR2:
+		// UniIC_DDR2_512M
+		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x00040608);
+		mmio_write_32(0x0004 + PHYD_BASE_ADDR, 0x02090B0C);
+		mmio_write_32(0x0008 + PHYD_BASE_ADDR, 0x0E01070D);
+		mmio_write_32(0x000C + PHYD_BASE_ADDR, 0x0A031605);
+		mmio_write_32(0x0010 + PHYD_BASE_ADDR, 0x10110F14);
+		mmio_write_32(0x0014 + PHYD_BASE_ADDR, 0x00121513);
+		mmio_write_32(0x0018 + PHYD_BASE_ADDR, 0x00000000);
+		mmio_write_32(0x001C + PHYD_BASE_ADDR, 0x00000100);
+		mmio_write_32(0x0020 + PHYD_BASE_ADDR, 0x61437025);
+		mmio_write_32(0x0024 + PHYD_BASE_ADDR, 0x00000008);
+		mmio_write_32(0x0028 + PHYD_BASE_ADDR, 0x63410725);
+		mmio_write_32(0x002C + PHYD_BASE_ADDR, 0x00000008);
+		break;
 	case DDR_VENDOR_ETRON_512M_DDR2:
 		// ETRON_DDR2_512
 		mmio_write_32(0x0000 + PHYD_BASE_ADDR, 0x070B090C);
diff --git a/drivers/ddr/ddr_pkg_info.c b/drivers/ddr/ddr_pkg_info.c
index d332e9c..4b8f5a8 100644
--- a/drivers/ddr/ddr_pkg_info.c
+++ b/drivers/ddr/ddr_pkg_info.c
@@ -78,7 +78,7 @@ void read_ddr_pkg_info(void)
 		ddr_type = DDR_TYPE_DDR3;
 		break;
 	case 0xb: //NY 1Gb DDR3
-		ddr_vendor = DDR_VENDOR_NY_1G;
+		ddr_vendor = DDR_VENDOR_NY_N20_1G;
 		ddr_capacity = DDR_CAPACITY_1G;
 		//pkg = PKG_QFN;
 		ddr_type = DDR_TYPE_DDR3;
diff --git a/include/drivers/ddr/ddr_pkg_info.h b/include/drivers/ddr/ddr_pkg_info.h
index c80b9f9..4288b28 100644
--- a/include/drivers/ddr/ddr_pkg_info.h
+++ b/include/drivers/ddr/ddr_pkg_info.h
@@ -13,7 +13,12 @@
 #define DDR_VENDOR_PM_1G		0b01000
 #define DDR_VENDOR_ETRON_512M_DDR2	0b01001
 #define DDR_VENDOR_ESMT_N25_1G		0b01010
-#define DDR_VENDOR_NY_1G		0b01011
+#define DDR_VENDOR_NY_N20_1G		0b01011
+#define DDR_VENDOR_ESMT_N21_2G		0b01100
+#define DDR_VENDOR_ESMT_N19_4G		0b01101
+#define DDR_VENDOR_UNILC_N25_512M_DDR2	0b01110
+#define DDR_VENDOR_UNILC_N25_1G		0b01111
+#define DDR_VENDOR_UNILC_N21_2G		0b10000
 
 #define DDR_CAPACITY_UNKNOWN		0b000
 #define DDR_CAPACITY_512M		0b001
-- 
2.25.1

